rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,318 @@
+Templated fields will be translated. Rather than reading from the specified 
+location, these will read from the runnable's inputs field. This can be done 
+simply by prepending the template paths with runnable.spec.inputs
+
+Example:
+```yaml
+kind: ClusterSourceTemplate    # <== Now a source template
+spec:
+   template:
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$
+```
+
+Will translate to a ClusterRunTemplate with:
+```yaml
+kind: ClusterRunTemplate
+spec:
+   template:
+         params:
+            - name: repository
+              value: $(runnable.spec.inputs.workload.spec.source.git.url)$
+```
+
+and a Runnable with
+```yaml
+kind: Runnable
+spec:
+   inputs:
+      workload.spec.source.git.url: *some-value-from-workload.spec.source.git.url*
+```
+
+The runnable and ClusterRunTemplate will be created with matching names and 
+a matching runTemplateRef. Assuming the example above and a workload with 
+the name `app`:
+
+```yaml
+---
+kind: Runnable
+metadata:
+   name: app-linter
+spec:
+   runTemplateRef:
+      name: app-linter
+---
+kind: ClusterRunTemplate
+metadata:
+   name: app-linter
+```
+
+The ClusterRunTemplate will specify outputs that match the output paths on 
+the template. Given the example above:
+
+```yaml
+---
+kind: ClusterRunTemplate
+spec:
+   outputs:
+      url: spec.params[?(@.name=="repository")].value
+      revision: spec.params[?(@.name=="revision")].value
+```
+
+When reading the runnable, Cartographer will look at the fields that match 
+the type of template. E.g. given a ClusterSourceTemplate with a 
+`runnable:true` field, Cartographer will always create a Runnable and always 
+read the runnable's .status.outputs.url field and .status.outputs.revision 
+field. Similarly for image, carto will read .status.outputs.image; for 
+config it will read .status.outputs.config.
+
+# Migration
+[migration]: #migration
+
+This is an additive change and does not present any breaking changes. Users 
+who wish to continue to hand create ClusterRunTemplates and template out 
+Runnables will be able to do so.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This RFC exchanges ease of usage for the app operator for complexity in the 
+controller. If we decide that app operators have no issue with creating 
+ClusterRunTemplates and templating Runnables, there is no need to undertake 
+this work.
+
+# Alternatives
+[alternatives]: #alternatives
+
+An alternative is to drop the Runnable object altogether. Cartographer could 
+directly manage the creation and reporting of objects 
+
+# Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and bad.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+## Resource reporting on workload
+The workload reports the resources created from each step. Should this point 
+to the Runnable object or to the most recently successful immutable object 
+created?
```

Comment:
I think pointing to the most recent successful object should be the behavior. This will match the object that the user has templated out. It will also save the user from going to the runnable and then inspecting the runnable to determine the object which is providing the values reflected there.

Reply:
Runnable/ClusterRunTemplates are currently hard-coded to treat Tekton's success condition as the evidence of success. When the templated object's `status.conditions[?(@.type=="Succeeded")].status == True` the templated object is considered successful.